### PR TITLE
Refactors residual damage

### PR DESCRIPTION
**Pull request by AlexOn1ine**
more readable and less prone to bugs

### CODE REVIEW

File: src/battle_util.c [hunk 1]
```diff
@@ -2936,14 +2929,31 @@ u8 DoBattlerEndTurnEffects(void)
...
             }
             gBattleStruct->turnEffectsTracker++;
             break;
+        case ENDTURN_GMAX_MOVE_RESIDUAL_DAMAGE:
+            {
+                u32 side = GetBattlerSide(gBattlerAttacker);
+                if (gSideTimers[side].damageNonTypesTimer
+                    && !IS_BATTLER_OF_TYPE(gBattlerAttacker, gSideTimers[side].damageNonTypesType)
+                    && IsBattlerAlive(gBattlerAttacker)
+                    && GetBattlerAbility(gBattlerAttacker) != ABILITY_MAGIC_GUARD)
```
Comment (Bassoonian) [inline src/battle_util.c:2946]:
```suggestion
                if (gSideTimers[side].damageNonTypesTimer
                 && !IS_BATTLER_OF_TYPE(gBattlerAttacker, gSideTimers[side].damageNonTypesType)
                 && IsBattlerAlive(gBattlerAttacker)
                 && GetBattlerAbility(gBattlerAttacker) != ABILITY_MAGIC_GUARD)
```

File: src/battle_util.c [hunk 2]
```diff
@@ -2936,14 +2929,31 @@ u8 DoBattlerEndTurnEffects(void)
         case ENDTURN_SEA_OF_FIRE_DAMAGE:
             if (IsBattlerAlive(battler) && gSideStatuses[GetBattlerSide(battler)] & SIDE_STATUS_SEA_OF_FIRE)
             {
-                gBattleMoveDamage = gBattleMons[battler].maxHP / 8;
+                gBattleMoveDamage = GetNonDynamaxMaxHP(battler) / 8;
```
Comment (AlexOn1ine) [inline src/battle_util.c:2932]:
```suggestion
                gBattleMoveDamage = gBattleMons[battler].maxHP  / 8;
```

File: src/battle_util.c [hunk 3]
```diff
@@ -2936,14 +2929,31 @@ u8 DoBattlerEndTurnEffects(void)
...
             break;
+        case ENDTURN_GMAX_MOVE_RESIDUAL_DAMAGE:
+            {
+                u32 side = GetBattlerSide(gBattlerAttacker);
+                if (gSideTimers[side].damageNonTypesTimer
+                    && !IS_BATTLER_OF_TYPE(gBattlerAttacker, gSideTimers[side].damageNonTypesType)
+                    && IsBattlerAlive(gBattlerAttacker)
+                    && GetBattlerAbility(gBattlerAttacker) != ABILITY_MAGIC_GUARD)
+                {
+                    gBattleMoveDamage = GetNonDynamaxMaxHP(battler) / 6;
```
Comment (AlexOn1ine) [inline src/battle_util.c:2948]:
```suggestion
                    gBattleMoveDamage = gBattleMons[battler].maxHP / 6;
```

File: src/battle_util.c [hunk 4]
```diff
@@ -2936,14 +2929,31 @@ u8 DoBattlerEndTurnEffects(void)
         case ENDTURN_SEA_OF_FIRE_DAMAGE:
             if (IsBattlerAlive(battler) && gSideStatuses[GetBattlerSide(battler)] & SIDE_STATUS_SEA_OF_FIRE)
             {
-                gBattleMoveDamage = gBattleMons[battler].maxHP / 8;
+                gBattleMoveDamage = gBattleMons[battler].maxHP  / 8;
```
Comment (AlexOn1ine) [inline src/battle_util.c:2932]:
```suggestion
                gBattleMoveDamage = gBattleMons[battler].maxHP / 8;
```

**Comment by AlexOn1ine** [comment]
Converting to draft because I noticed more scripts I would like to remove

**Comment by AlexOn1ine** [comment]
Ready for review!

I also refactored weather damage and added a simple animation for Ice Body heal that can be used anywhere else.

File: test/battle/weather/hail.c [hunk 5]
```diff
@@ -53,3 +53,20 @@ SINGLE_BATTLE_TEST("Hail fails if Desolate Land or Primordial Sea are active")
         }
     }
 }
+
+DOUBLE_BATTLE_TEST("Hail deals based on turn order")
```
Comment (AlexOn1ine) [inline test/battle/weather/hail.c:57]:
```suggestion
DOUBLE_BATTLE_TEST("Hail deals damage based on turn order")
```

File: src/battle_script_commands.c [hunk 6]
```diff
@@ -749,7 +749,7 @@ void (* const gBattleScriptingCommandsTable[])(void) =
     Cmd_tryKO,                                   //0x93
     Cmd_damagetohalftargethp,                    //0x94
     Cmd_unused_95,                               //0x95
-    Cmd_weatherdamage,                           //0x96
+    Cmd_unused_96,                             //0x96
     Cmd_tryinfatuating,                          //0x97
```
Comment (Bassoonian) [inline src/battle_script_commands.c:753]:
```suggestion
    Cmd_unused_96,                               //0x96
    Cmd_tryinfatuating,                          //0x97
```

File: test/battle/weather/hail.c [hunk 7]
```diff
@@ -53,3 +53,20 @@ SINGLE_BATTLE_TEST("Hail fails if Desolate Land or Primordial Sea are active")
         }
     }
 }
+
+DOUBLE_BATTLE_TEST("Hail deals damage based on turn order")
```
Comment (Bassoonian) [inline test/battle/weather/hail.c:57]: Should we have analogous tests for Sandstorm?

Comment (AlexOn1ine) [reply test/battle/weather/hail.c:57]: done

**Review by Bassoonian — APPROVED** [review]

**Pull request merged**